lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,201 @@
+            type: string
+            env:
+                - name: TOWER_HOST
+            required: True
+        tower_username:
+            description: The user that you plan to use to access inventories on tower.
+            type: string
+            env:
+                - name: TOWER_USERNAME
+            required: True
+        tower_password:
+            description: The password for tower_username.
+            type: string
+            env:
+                - name: TOWER_PASSWORD
+            required: True
+        tower_inventory:
+            description: The ID of the tower inventory you wish to import.
+            type: string
+            env:
+                - name: TOWER_INVENTORY
+            required: True
+        tower_verify_ssl:
+            description: Specify whether Ansible should verify the SSL certificate of Ansible tower.
+            type: bool
+            default: True
+            env:
+                - name: TOWER_VERIFY_SSL
+            required: False
+'''
+
+EXAMPLES = '''
+# Before you execute the following commands, you should make sure this file is in your plugin path,
+# and you enabled this plugin.
+
+# Example for using tower_inventory.yml file
+
+plugin: tower
+tower_host: your_tower_server_network_address
+tower_username: your_username
+tower_password: your_password
+tower_inventory: the_ID_of_targeted_inventory
+# Then you can run the following command.
+# If some of the arguments are missing, Ansible will attempt to read them from environment variables.
+# ansible-inventory -i /path/to/tower_inventory.yml --list
+
+# Example for reading from environment variables:
+
+# Set environment variables:
+# export TOWER_HOST=YOUR_TOWER_HOST_ADDRESS
+# export TOWER_USERNAME=YOUR_TOWER_USERNAME
+# export TOWER_PASSWORD=YOUR_TOWER_PASSWORD
+# export TOWER_INVENTORY=THE_ID_OF_TARGETED_INVENTORY
+# Read the inventory specified in TOWER_INVENTORY from Ansible Tower, and list them.
+# The inventory path must always be tower_inventory if you are reading all settings from environment variables.
+# ansible-inventory -i tower_inventory --list
+'''
+
+import re
+import os
+from ansible.module_utils import six
+from ansible.module_utils._text import to_text
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.plugins.inventory import BaseInventoryPlugin
+
+try:
+    from urlparse import urljoin
+except ImportError:
+    from urllib.parse import urljoin
+
+try:
+    import requests
+    from requests.auth import HTTPBasicAuth
+except ImportError:
+    raise AnsibleError('You have to install python requests library to use tower inventory plugin.')
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'tower'
+    # Stays backward compatible with tower inventory script.
+    # If the user supplies 'tower_inventory' as path, the plugin will read from environment variables.
+    read_all_configs_from_env_var = False
```

Comment:
why force either/or? you should be able to read some things from the config file, others from env vars, user might not wan't all eggs in one basket

Reply:
It doesn't force either/or. Just indicates when the plugin needs to call set_options (or have self._read_config_data do it) if the inventory file doesn't exist.

Reply:
direct only gets used for inventory plugins if there's a config file. Otherwise get_option suffices.

Reply:
@s-hertel Thanks for the clarification! That is my intention.